### PR TITLE
fix(storage): make ordered merge iter cancellation safe

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Our current implementation of the ordered merge iter will take all nodes with the same key as the top key out, and then advance all the nodes, and then put the nodes back to the heap. Advancing the node iterators is async, and may gets cancelled anytime. While the nodes are taken out from the heap, the merge iterator is in an inconsistent state, and will panic when trying to get the key, as https://github.com/risingwavelabs/risingwave/issues/6637 does.

The panic in https://github.com/risingwavelabs/risingwave/issues/6637 can be reproduced when we construct the following iterator
```
let mut merge_iter = UnorderedMergeIteratorInner::new(vec![
            OrderedMergeIteratorInner::new(once(CancellationTestIterator {})),
            OrderedMergeIteratorInner::new(once(CancellationTestIterator {})),
        ]);
```
, where the `CancellationTestIterator` always wait on pending. When we drive the future returned from `next` and then drop the future, we will panic.

In this PR, we fix this issue by not taking the node out from the heap. At the beginning, we will save the key of the top node. And then we will loop on calling `next` on the top node if the current key of the current top node is the same as the top node key at the beginning. To reduce the number of memory allocation, we add a new Vec<u8> field in the struct of merge iterator to store the top table key.

We add a new unit test to drive the test merge iterator mentioned above. The current main branch will fail in the unit test, while this branch will pass.

In terms of performance, we even get great performance improvement in the ordered merge iterator.
The main branch
```
bench-merge-iter/unordered
                        time:   [171.77 ms 172.11 ms 172.48 ms]
bench-merge-iter/ordered
                        time:   [228.34 ms 228.68 ms 229.02 ms]
bench-enum-merge-iter/unordered
                        time:   [213.87 ms 214.17 ms 214.48 ms]
bench-enum-merge-iter/ordered
                        time:   [270.49 ms 270.84 ms 271.18 ms]
```

Current branch without Vec<u8> buffer
```
bench-merge-iter/unordered
                        time:   [175.76 ms 175.95 ms 176.13 ms]
bench-merge-iter/ordered
                        time:   [204.35 ms 204.62 ms 204.89 ms]
bench-enum-merge-iter/unordered
                        time:   [213.93 ms 214.24 ms 214.55 ms]
bench-enum-merge-iter/ordered
                        time:   [240.33 ms 240.69 ms 241.04 ms]
```

Current branch with Vec<u8> buffer
```
bench-merge-iter/unordered
                        time:   [172.57 ms 172.86 ms 173.12 ms]
bench-merge-iter/ordered
                        time:   [182.48 ms 182.75 ms 183.01 ms]
bench-enum-merge-iter/unordered
                        time:   [213.55 ms 213.88 ms 214.21 ms]
bench-enum-merge-iter/ordered
                        time:   [216.68 ms 217.03 ms 217.39 ms]

```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
fix https://github.com/risingwavelabs/risingwave/issues/6637

### CODE REVIEW

**Review by hzxa21 — APPROVED** [review]
LGTM! Nice catch. This is really subtle.

I believe the performance improvement may come from the fact that we won't need to pop all nodes with the same table_key and re-push them to the heap.

File: src/storage/src/hummock/iterator/merge_inner.rs [hunk 1]
```diff
@@ -202,41 +205,43 @@ impl<I: HummockIterator> MergeIteratorNext for OrderedMergeIteratorInner<I> {
...
-            // Put the popped nodes back to the heap if valid or unused_iters if invalid.
-            for mut node in popped_nodes {
-                match node.iter.next().await {
-                    Ok(_) => {}
-                    Err(e) => {
-                        // If the iterator returns error, we should clear the heap, so that this
-                        // iterator becomes invalid.
+            };
+            loop {
+                // We don't use `while let` or `if let` here because the current borrow checker is
```
Comment (BugenZhao) [inline src/storage/src/hummock/iterator/merge_inner.rs:222]: Does `let Some(xx) = xxx else { break }` work here?

**Review by BugenZhao — APPROVED** [review]
LGTM! Thanks for the quick fix!

File: src/storage/src/hummock/iterator/merge_inner.rs (hunk 1 above)
Comment (wenym1) [reply src/storage/src/hummock/iterator/merge_inner.rs]: Thanks. It does work. I have fixed it.

**Pull request merged**